src/core/proto/netlink_socket_mgr.cpp

```diff
@@ -32,173 +32,72 @@
-// Return length of received data.
-// Parameters:
-//      sockfd  : netlink socket
-//      pid     : expected opaque pid value
-//      seq     : expected opaque seq value
-//      buf     : [out] read reply
-int netlink_socket_mgr::recv_info(int sockfd, uint32_t pid, uint32_t seq, char *buf)
-{
-    struct nlmsghdr *nlHdr;
-    int readLen;
-    int msgLen = 0;
-    char *buf_ptr = buf;
-
-    do {
-        // Receive response from the kernel
-        BULLSEYE_EXCLUDE_BLOCK_START
-        if ((readLen = SYSCALL(recv, sockfd, buf_ptr, MSG_BUFF_SIZE - msgLen, 0)) < 0) {
-            __log_err("NL socket read failed, errno = %d", errno);
-            return -1;
-        }
-
-        nlHdr = (struct nlmsghdr *)buf_ptr;
-
-        // Check if the header is valid
-        if ((NLMSG_OK(nlHdr, (u_int)readLen) == 0) || (nlHdr->nlmsg_type == NLMSG_ERROR)) {
-            __log_err("Error in received packet, readLen = %d, msgLen = %d, type=%d, bufLen = %d",
-                      readLen, nlHdr->nlmsg_len, nlHdr->nlmsg_type, MSG_BUFF_SIZE);
-            if ((int)nlHdr->nlmsg_len >= MSG_BUFF_SIZE - msgLen) {
-                __log_err("The buffer we pass to netlink is too small for reading the whole table");
-            }
-            return -1;
-        }
-        BULLSEYE_EXCLUDE_BLOCK_END
-
-        if ((nlHdr->nlmsg_seq != seq) || (nlHdr->nlmsg_pid != pid)) {
-            // Skip not expected messages
-            continue;
-        }
-
-        buf_ptr += readLen;
-        msgLen += readLen;
-
-        // Loop until this is the last message of expected reply
-    } while (nlHdr->nlmsg_type != NLMSG_DONE && (nlHdr->nlmsg_flags & NLM_F_MULTI));
-
-    return msgLen;
-}
+    if (nl_connect(sockfd, NETLINK_ROUTE) < 0) {
+        __log_err("NL socket Connection: ");
+        nl_socket_free(sockfd);
+        throw_xlio_exception("Failed nl_connect");
+    }
 
-// Update data in a table
-void netlink_socket_mgr::update_tbl(nl_data_t data_type)
-{
-    struct nlmsghdr *nl_msg = nullptr;
-    char *buf;
-    int len = 0;
+    struct nl_cache *cache_state = {0};
+    int err = 0;
 
-    // Opaque netlink information
-    uint32_t nl_pid = getpid();
-    uint32_t nl_seq = static_cast<uint32_t>(data_type);
+    // cache allocation fetches the latest existing rules/routes
+    if (data_type == RULE_DATA_TYPE) {
 
-    __log_dbg("");
+        err = rtnl_rule_alloc_cache(sockfd, AF_INET, &cache_state);
+    } else if (data_type == ROUTE_DATA_TYPE) {
 
-    buf = new char[MSG_BUFF_SIZE];
-    if (!buf) {
-        __log_err("NL message buffer allocation failed");
-        return;
+        err = rtnl_route_alloc_cache(sockfd, AF_INET, 0, &cache_state);
     }
 
-    build_request(data_type, nl_pid, nl_seq, buf, &nl_msg);
-    if (query(nl_msg, buf, len)) {
-        parse_tbl(buf, len);
+    if (err < 0) {
+        throw_xlio_exception("Failed to allocate route cache");
     }
 
-    delete[] buf;
-    __log_dbg("Done");
+    parse_tbl(cache_state);
 }
```

Comment:
the cache needs to be destroyed at the end.

Reply:
fixed